File: src/providers/integraicp/schemas.ts
import { z } from "zod";

// Contract taken from docs/vendor/integraicp-api-reference-v3.md
export const authenticationsQuerySchema = z.object({
  subject_key: z.string().optional(),
  subject_type: z.string().optional(),
  secret_data: z.string(),
  secret_type: z.string().optional(),
  callback_uri: z.string().url(),
  autostart: z
    .enum(["true", "false"])
    .optional()
    .transform((v) => v === "true"),
  credential_lifetime: z.coerce.number().optional(),
  clearance_lifetime: z.coerce.number().optional(),
});

export const credentialsQuerySchema = z.object({
  secret_data: z.string(),
  secret_type: z.string().optional(),
});

export const credentialsParamsSchema = z.object({
  channelId: z.string(),
  credentialId: z.string(),
});

const signatureRequestItemSchema = z.object({
  contentId: z.string().optional(),
  contentDigest: z.string(),
  contentDescription: z.string().optional(),
  signaturePolicy: z.enum(["RAW", "CMS"]).optional(),
});

export const signaturesBodySchema = z.object({
  credentialId: z.string(),
  secretType: z.string().optional(),
  secretData: z.string(),
  requests: z.array(signatureRequestItemSchema).min(1),
});

export type AuthenticationsQuery = z.infer<typeof authenticationsQuerySchema>;
export type CredentialsQuery = z.infer<typeof credentialsQuerySchema>;
export type SignaturesBody = z.infer<typeof signaturesBodySchema>;

File: internal/providers/integraicp/schemas.go
package integraicp

import (
	"net/url"
	"strconv"
)

// Contract taken from docs/vendor/integraicp-api-reference-v3.md.

type AuthenticationsQuery struct {
	SubjectKey        *string
	SubjectType       *string
	SecretData        string
	CallbackURI       string
	Autostart         bool
	ClearanceLifetime *int64
}

func parseAuthenticationsQuery(q url.Values) (*AuthenticationsQuery, bool) {
	secretData := q.Get("secret_data")
	callbackURI := q.Get("callback_uri")
	if secretData == "" || callbackURI == "" {
		return nil, false
	}
	if _, err := url.ParseRequestURI(callbackURI); err != nil {
		return nil, false
	}

	out := &AuthenticationsQuery{SecretData: secretData, CallbackURI: callbackURI}

	if v := q.Get("subject_key"); v != "" {
		out.SubjectKey = &v
	}
	if v := q.Get("subject_type"); v != "" {
		out.SubjectType = &v
	}

	if v := q.Get("autostart"); v != "" {
		if v != "true" && v != "false" {
			return nil, false
		}
		out.Autostart = v == "true"
	}

	if v := q.Get("clearance_lifetime"); v != "" {
		n, err := strconv.ParseInt(v, 10, 64)
		if err != nil {
			return nil, false
		}
		out.ClearanceLifetime = &n
	}

	return out, true
}

type CredentialsQuery struct {
	SecretData string
}

func parseCredentialsQuery(q url.Values) (*CredentialsQuery, bool) {
	secretData := q.Get("secret_data")
	if secretData == "" {
		return nil, false
	}
	return &CredentialsQuery{SecretData: secretData}, true
}

type SignatureRequestItem struct {
	ContentID          *string `json:"contentId,omitempty"`
	ContentDigest      string  `json:"contentDigest"`
	ContentDescription *string `json:"contentDescription,omitempty"`
}

type SignaturesBody struct {
	CredentialID string                  `json:"credentialId"`
	SecretData   string                  `json:"secretData"`
	Requests     []SignatureRequestItem `json:"requests"`
}

func (b SignaturesBody) valid() bool {
	return b.CredentialID != "" && b.SecretData != "" && len(b.Requests) >= 1
}
